### PR TITLE
Fix pull request update showing too many commits with multiple branches

### PR DESCRIPTION
**Pull request by brechtvl**
When the base repository contains multiple branches with the same commits as the base branch, pull requests can show a long list of commits already in the base branch as having been added.

What this is supposed to do is exclude commits already in the base branch. But the mechansim to do so assumed a commit only exists in a single branch. Now use `git rev-list A B --not branchName` instead of filtering commits afterwards.

The logic to detect if there was a force push also was wrong for multiple branches. If the old commit existed in any branch in the base repository it would assume there was no force push. Instead check if the old commit is an ancestor of the new commit.

### CODE REVIEW

**Comment by brechtvl** [comment]
An example of this problem can be seen here:
https://projects.blender.org/blender/blender/pulls/104556

File: modules/git/repo_commit.go [hunk 1]
```diff
@@ -323,6 +323,28 @@ func (repo *Repository) CommitsBetweenLimit(last, before *Commit, limit, skip in
...
+	var stdout []byte
+	var err error
+	if before == nil {
+		stdout, _, err = NewCommand(repo.Ctx, "rev-list").AddDynamicArguments(last.ID.String()).AddArguments("--not").AddDynamicArguments(baseBranch).RunStdBytes(&RunOpts{Dir: repo.Path})
+	} else {
+		stdout, _, err = NewCommand(repo.Ctx, "rev-list").AddDynamicArguments(before.ID.String() + ".." + last.ID.String()).AddArguments("--not").AddDynamicArguments(baseBranch).RunStdBytes(&RunOpts{Dir: repo.Path})
+		if err != nil && strings.Contains(err.Error(), "no merge base") {
+			// future versions of git >= 2.28 are likely to return an error if before and last have become unrelated.
+			// previously it would return the results of git rev-list before last so let's try that...
+			stdout, _, err = NewCommand(repo.Ctx, "rev-list").AddDynamicArguments(before.ID.String(), last.ID.String()).AddArguments("--not").AddDynamicArguments(baseBranch).RunStdBytes(&RunOpts{Dir: repo.Path})
```
Comment (wxiaoguang) [inline modules/git/repo_commit.go]: I haven't looked into detaills, a small hint: `.AddArguments("--not").AddDynamicArguments(baseBranch)` could be rewritten to `.AddOptionValues("--not", baseBranch)` for short. 😁

File: modules/git/repo_commit.go [hunk 2]
```diff
@@ -323,6 +323,28 @@ func (repo *Repository) CommitsBetweenLimit(last, before *Commit, limit, skip in
...
+	var stdout []byte
+	var err error
+	if before == nil {
+		stdout, _, err = NewCommand(repo.Ctx, "rev-list").AddDynamicArguments(last.ID.String()).AddOptionValues("--not", baseBranch).RunStdBytes(&RunOpts{Dir: repo.Path})
+	} else {
+		stdout, _, err = NewCommand(repo.Ctx, "rev-list").AddDynamicArguments(before.ID.String()+".."+last.ID.String()).AddOptionValues("--not", baseBranch).RunStdBytes(&RunOpts{Dir: repo.Path})
+		if err != nil && strings.Contains(err.Error(), "no merge base") {
+			// future versions of git >= 2.28 are likely to return an error if before and last have become unrelated.
+			// previously it would return the results of git rev-list before last so let's try that...
+			stdout, _, err = NewCommand(repo.Ctx, "rev-list").AddDynamicArguments(before.ID.String(), last.ID.String()).AddOptionValues("--not", baseBranch).RunStdBytes(&RunOpts{Dir: repo.Path})
```
Comment (delvh) [inline modules/git/repo_commit.go:338]: I don't quite understand, the only difference is that `before..last` has been converted to `before last`?
How does that help?

**Review by delvh — APPROVED** [review]

File: modules/git/repo_commit.go (hunk 2 above)
Comment (brechtvl) [reply modules/git/repo_commit.go:338]: I don't know either, this just copies the logic from `CommitsBetween` and `CommitsBetweenLimit`.

**Comment by lunny** [comment]
We have a function named `IsForcePush` in `modules/repository/push.go`. Looks like they are different git commands. But both are work?

**Comment by brechtvl** [comment]
From what I can tell the commands are equivalent, with `merge-base --is-ancestor` being a more recent and convenient addition.

`IsForcePush` only affects trace logging currently, so not much risk breaking anything there. So I deduplicated and simplified the code.

File: modules/git/repo_commit.go [hunk 3]
```diff
@@ -323,6 +323,27 @@ func (repo *Repository) CommitsBetweenLimit(last, before *Commit, limit, skip in
 	return repo.parsePrettyFormatLogToList(bytes.TrimSpace(stdout))
 }
 
+// CommitsBetween returns a list that contains commits between [before, last), excluding commits in baseBranch.
```
Comment (lunny) [inline modules/git/repo_commit.go]:
```suggestion
// CommitsBetweenNotBase returns a list that contains commits between [before, last), excluding commits in baseBranch.
```

**Review by lunny — APPROVED** [review]

**Comment by jolheiser** [comment]
🎺 🤖

**Pull request merged**